src/main/java/address/sync/CloudSimulator.java

```diff
@@ -1,129 +1,670 @@
+                CloudPerson person = new CloudPerson(java.util.UUID.randomUUID().toString(),
+                                                     java.util.UUID.randomUUID().toString());
+                System.out.println("Cloud simulator: adding " + person);
+                personList.add(person);
             }
-            modifiedData.add(person);
         }
+    }
+
+    private void addCloudTagsBasedOnChance(List<CloudTag> tagList) {
+        for (int i = 0; i < MAX_NUM_PERSONS_TO_ADD; i++) {
+            if (RANDOM_GENERATOR.nextDouble() <= ADD_TAG_PROBABILITY) {
+                CloudTag tag = new CloudTag(java.util.UUID.randomUUID().toString());
+                System.out.println("Cloud simulator: adding tag '" + tag + "'");
+                tagList.add(tag);
+            }
+        }
+    }
+
+    private void modifyCloudPersonBasedOnChance(CloudPerson cloudPerson) {
+        if (RANDOM_GENERATOR.nextDouble() > MODIFY_PERSON_PROBABILITY) return;
+        System.out.println("Cloud simulator: modifying person '" + cloudPerson + "'");
+        cloudPerson.setCity(java.util.UUID.randomUUID().toString());
+        cloudPerson.setStreet(java.util.UUID.randomUUID().toString());
+        cloudPerson.setPostalCode(String.valueOf(RANDOM_GENERATOR.nextInt(999999)));
+    }
+
+    private void modifyCloudTagBasedOnChance(CloudTag cloudTag) {
+        if (RANDOM_GENERATOR.nextDouble() > MODIFY_TAG_PROBABILITY) return;
+        System.out.println("Cloud simulator: modifying tag '" + cloudTag + "'");
+        cloudTag.setName(UUID.randomUUID().toString());
+    }
+
+    private void delayRandomAmount() {
+        long delayAmount = RANDOM_GENERATOR.nextInt(DELAY_RANGE) + MIN_DELAY_IN_SEC;
+        try {
+            TimeUnit.SECONDS.sleep(delayAmount);
+        } catch (InterruptedException e) {
+            System.out.println("Error occurred while adding cloud response delay.");
+        }
+    }
+
+    private ByteArrayInputStream convertToInputStream(Object object) {
+        try {
+            return new ByteArrayInputStream(JsonUtil.toJsonString(object).getBytes());
+        } catch (JsonProcessingException e) {
+            e.printStackTrace();
+            return null;
+        }
+    }
+
+    private int getNumberOfRequestsRequired(int dataSize, int resourcesPerPage) {
+        return (int) Math.ceil((double) dataSize / resourcesPerPage);
+    }
+
+    private void deletePersonFromData(List<CloudPerson> personList, String firstName, String lastName)
+            throws NoSuchElementException {
+        CloudPerson deletedPerson = getPersonIfExists(personList, firstName, lastName);
+        deletedPerson.setDeleted(true);
+    }
+
+    private CloudTag addTag(List<CloudTag> tagList, CloudTag newTag) {
+        if (newTag == null) throw new IllegalArgumentException("Tag cannot be null");
+        String tagName = newTag.getName();
+        if (tagName == null) throw new IllegalArgumentException("Fields cannot be null");
```

Comment:
I think we can use `newTag.getName() == null` here since `tagName` not used anywhere else.